internal/manager/manager.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package manager contains nsmgr main code.
+package manager
+
+import (
+	"context"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/networkservicemesh/cmd-nsmgr/internal/authz"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/nsmgr"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/authorize"
+	"github.com/networkservicemesh/sdk/pkg/tools/callback"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"github.com/networkservicemesh/cmd-nsmgr/internal/config"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+	"github.com/networkservicemesh/sdk/pkg/tools/signalctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/spanhelper"
+	"github.com/networkservicemesh/sdk/pkg/tools/spiffejwt"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffetls/tlsconfig"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+const tcpSchema = "tcp"
+
+type manager struct {
+	ctx           context.Context
+	configuration *config.Config
+	span          spanhelper.SpanHelper
+	cancelFunc    context.CancelFunc
+	registryCC    *grpc.ClientConn
+	mgr           nsmgr.Nsmgr
+	source        *workloadapi.X509Source
+	svid          *x509svid.SVID
+	server        *grpc.Server
+}
+
+func (m *manager) Stop() {
+	m.cancelFunc()
+	m.server.Stop()
+	_ = m.source.Close()
+}
+
+func (m *manager) initSecurity() (err error) {
+	// Get a X509Source
+	m.source, err = workloadapi.NewX509Source(m.ctx)
+	if err != nil {
+		logrus.Fatalf("error getting x509 source: %+v", err)
+	}
+	m.svid, err = m.source.GetX509SVID()
+	if err != nil {
+		logrus.Fatalf("error getting x509 svid: %+v", err)
+	}
+	logrus.Infof("SVID: %q", m.svid.ID)
+	return
+}
+
+// RunNsmgr - start nsmgr.
+func RunNsmgr(ctx context.Context, configuration *config.Config) error {
+	starttime := time.Now()
+
+	m := &manager{
+		configuration: configuration,
+		span:          spanhelper.FromContext(ctx, "start"),
+	}
+
+	// Update context
+	m.ctx = signalctx.WithSignals(m.span.Context())
```

Comment:
You *really* need to do this near the very very top of main(), otherwise you can miss signals.

Reply:
Good catch, actually main had it as first line, and this one call is not required :-)